### ISSUE TITLE
When moving to parse server - application bug of infinite loop causes 100% CPU crashes

### ISSUE THREAD

**Issue by ronnno** [ParsePlatform/parse-server#2411]
Application bug caused CPU to reach 100% in some rare cases.
Since in Parse service we were ignorant of these conditions the same code would run, break, and freed by the process timeout mechanism in Parse.

In Parse-server we are suddenly allowed to consume as much CPU as we need (a good thing) but it exposes this bug.
This is not a real parse server issue - it was fixed in our application.

**Comment by ronnno** [comment]
Another thing that might be relevant. Until now I was using serverURL with the public address to my server. So REST API calls from cloud code were routed through the load balancer and could reach another node. some calls may do that several times in cascading triggers (beforeSave and afterSave). so it's possible we might have had some reentrant locking issue?  I'm just guessing here but in case this can help explain the crashes I'm just putting it here as well.

after asking about that in issue #2412 I now use localhost for my serverURL.

**Comment by ronnno** [comment]
This issue was a bug in one of our cloud code functions that caused an infinite loop.
It was always there - but since Parse.com would not report these issues and just cut the process after 15 seconds of processing we never noticed that in very rare cases the process failed to complete.

Since we moved to our own parse server - there is no process cut off after 15 seconds so the entire server stopped as a result of this bug. we fixed it only after learning how to recreate it on a dev env and then finding the cause was easy using a debugger.
 
if you don't have cloud code functions or triggers or scheduled jobs running or any javascript code of your own it's hard for me to guess what the reason might be. I would verify that the version you are running is current and create a new issue for this i guess.

And also try to rule out issues with garbage collection as I hear that if set up incorrectly this can cause process stalls. but to my knowledge should eventually free up so if it doesn't ever return to work its not a garbage collection issue.

Hope this helps

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.1.0-alpha.13](https://github.com/parse-community/parse-server/releases/tag/6.1.0-alpha.13)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.3.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/6.3.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.3.0-alpha.1](https://github.com/parse-community/parse-server/releases/tag/6.3.0-alpha.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.3.0](https://github.com/parse-community/parse-server/releases/tag/6.3.0)
